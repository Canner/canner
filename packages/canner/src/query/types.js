// @flow

export type QueryObject = null | {
  fields?: {
    [string]: QueryObject
  },
  args?: {
    [string]: any,
  },
  isPlural?: boolean,
  connection?: boolean,
  alias?: string
}