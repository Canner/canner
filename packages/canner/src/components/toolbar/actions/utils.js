import {
  forEach,
} from 'lodash';

export function flattenItems(items) {
  const fields = [];

  function loop(item, keyName = '') {
    let newKeyName = keyName;
    if (item.keyName) {
      newKeyName = `${newKeyName}${newKeyName ? '.' : ''}${item.keyName}`;
    }

    if (item.items && !item.items.type) {
      forEach(item.items, (item) => {
        loop(item, newKeyName);
      });
    } else if (item.items && item.items.type && item.items.items) {
      forEach(item.items.items, (item) => {
        loop(item, newKeyName);
      });
    } else {
      fields.push({
        ...item,
        keyName: newKeyName,
      });
    }
  }

  forEach(items, (item) => {
    loop(item, '');
  });

  return fields;
}
