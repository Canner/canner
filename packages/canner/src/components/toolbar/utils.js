// @flow

export function findAlwaysDisplayedFilterIndexes(filters: Array<Object>): Array<number> {
  const result = [];
  filters.forEach((filter, index) => {
    if (filter.alwaysDisplay) {
      result.push(index);
    }
  });
  return result;
}