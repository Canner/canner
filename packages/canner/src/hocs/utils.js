// @flow
import {mapValues, get, isPlainObject, isArray} from 'lodash';

export function getValue(value: any, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (isPlainObject(result)) {
      if ('edges' in result && 'pageInfo' in result) {
        return get(result, ['edges', key, 'node']);
      }
      return get(result, key);
    } else if (isArray(result)) {
      return get(result, key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: any) {
  if (isPlainObject(value)) {
    if (value.edges && value.pageInfo) {
      return value.edges.map(edge => parseConnectionToNormal(edge.node));
    }
    return mapValues(value, item => parseConnectionToNormal(item));
  } else if (isArray(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

export function defaultValue(type: string, relation: any) {
  switch (type) {
    case 'connection': {
      return {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false
        }
      }
    }
    case 'array': {
      return [];
    }
    case 'object': {
      return {};
    }
    case 'boolean': {
      return false;
    }
    case 'number': {
      return 0;
    }
    case 'string': {
      return '';
    }
    case 'relation': {
      if (relation.type === 'toMany') {
        return {
          edges: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          }
        };
      } else {
        return null;
      }
    }
    case 'image':
    case 'file': {
      return {
        url: '',
        contentType: '',
        name: '',
        size: 0,
        __typename: null
      }
    }
    case 'geoPoint': {
      return {
        placeId: '',
        address: '',
        lat: 122,
        lng: 23
      };
    }
    default: {
      return null;
    }
  }
}

export function paginate(rootValue: Object, field: string, currentPage: number, pageSize: number = 10) {
  const copyValue = JSON.parse(JSON.stringify(rootValue));
  const index = (currentPage - 1) * pageSize;
  copyValue[field].edges = copyValue[field].edges.slice(index, index + pageSize);
  return copyValue;
}

export function filterByWhere(rootValue: Object, field: string, where: Object) {
  const copyValue = JSON.parse(JSON.stringify(rootValue));
  copyValue[field].edges = copyValue[field].edges.filter(edge => {
    return Object.keys(where).reduce((result, key) => {
      const condition = where[key];
      const data = edge.node[key];
      return result && passCondition(data , condition);
    }, true);
  });
  return copyValue;
}

export function passCondition(data: any, condition: Object) {
  return Object.keys(condition).reduce((result, conditionKey) => {
    let isPass = true;
    const conditionValue = condition[conditionKey];
    switch(conditionKey) {
      case 'contains':
        isPass = data.indexOf(conditionValue) !== -1;
        break;
      case 'eq':
        isPass = data === conditionValue;
        break;
      case 'gt':
        isPass = data > conditionValue;
        break;
      case 'gte':
        isPass = data >= conditionValue;
        break;
      case 'lt':
        isPass = data < conditionValue;
        break;
      case 'lte':
        isPass = data <= conditionValue;
        break;
    }
    return result && isPass;
  }, true);
}