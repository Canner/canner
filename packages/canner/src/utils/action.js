// @flow
import {generateAction} from '../action';
import {update, merge, isPlainObject} from 'lodash';
import {createEmptyData} from 'canner-helpers';

export function createAction({
  relation,
  id,
  type,
  value,
  config,
  rootValue,
  items,
  transformGqlPayload
}: { 
  relation: any,
  id: {firstId: string, secondId: string} | string,
  type: 'create' | 'update' | 'delete' | 'swap',
  value: any,
  config: any,
  rootValue: any,
  value: any,
  items: any,
  transformGqlPayload?: Function
}) {
  if (type === 'create') {
    if (!config) {
      const emptyData = createEmptyData(items);
      if (isPlainObject(emptyData)) {
        value = merge(emptyData, value);
      } else {
        value = emptyData;
      }
    }
    if (typeof id === 'string' && id.split('/').length === 1) {
      update(value, 'id', id => id || randomId());
    }
  }
  return generateAction({
    id,
    updateType: type,
    value,
    rootValue,
    relation,
    transformGqlPayload
  });
}

function randomId() {
  return Math.random().toString(36).substr(2, 12);
}
