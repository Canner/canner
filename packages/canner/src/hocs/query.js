// @flow

import * as React from 'react';
import {List} from 'react-content-loader';
import Toolbar from '../components/toolbar/index';
import {Icon, Spin} from 'antd';
import {mapValues, isNil} from 'lodash';
import {getValue, defaultValue} from './utils';
import {withApollo} from 'react-apollo';
import type {HOCProps} from './types';
type State = {
  value: any,
  rootValue: any,
  originRootValue: any,
  isFetching: boolean,
}

@withApollo
export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<HOCProps, State> {
    key: string;
    subscription: any;

    constructor(props: HOCProps) {
      super(props);
      this.state = {
        value: null,
        rootValue: null,
        originRootValue: null,
        isFetching: true
      };
      this.key = props.refId.getPathArr()[0];
    }

    componentDidMount() {
      this.queryData().then(() => {
        this.subscribe();
      });
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    componentDidUpdate(prevProps: HOCProps) {
      const {refId} = prevProps;
      if (refId.toString() !== this.props.refId.toString()) {
        // refetch when route change
        this.unsubscribe();
        this.queryData(this.props).then(() => {
          this.subscribe();
        });
      }
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = () => {
      const {subscribe} = this.props;
      const subscription = subscribe(this.key, result => {
        const {data, rootValue} = result;
        this.updateData(data, rootValue);
      });
      this.subscription = subscription;
    }

    queryData = (props?: HOCProps): Promise<*> => {
      const {fetch} = props || this.props;
      this.setState({
        isFetching: true
      });
      return fetch(this.key).then(result => {
        const {data, rootValue} = result;
        this.updateData(data, rootValue);
      });
    }

    updateData = (data: any, rootValue: any) => {
      const {refId} = this.props;
      const newValue = getValue(data, refId.getPathArr());
      this.setState({
        originRootValue: data,
        rootValue,
        value: newValue,
        isFetching: false
      });
    }

    updateQuery = (paths: Array<string>, args: Object) => {
      const {updateQuery} = this.props;
      updateQuery(paths, args)
        .then(rewatch => {
          if (rewatch) {
            this.unsubscribe();
            this.queryData();
            this.subscribe();
          } else {
            this.queryData();
          }
        });
    }

    render() {
      const {value, isFetching, rootValue, originRootValue} = this.state;
      const {toolbar, query, refId, items, type, path, relation, pattern, keyName, request, deploy, nullable} = this.props;
      if (!originRootValue) {
        return <List style={{maxWidth: '600px'}}/>;
      }
      if (pattern === 'array') {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v.substr(1)]);

        return (
          <Toolbar items={items}
            toolbar={toolbar}
            args={args}
            query={query}
            refId={refId}
            keyName={keyName}
            originRootValue={originRootValue}
            rootValue={rootValue}
            getValue={getValue}
            updateQuery={this.updateQuery}
            request={request}
            deploy={deploy}
          >
            <SpinWrapper isFetching={isFetching}>
              <Com {...this.props} />
            </SpinWrapper>
          </Toolbar>
        );
      } else if (type === 'relation' && relation.type === 'toOne') {
        return (
            <Com {...this.props} showPagination={true} rootValue={rootValue} value={(value && value.id) ? value : defaultValue(type, relation)} />
        );
      } else if (type === 'relation' && relation.type === 'toMany') {
        return (
            <Com {...this.props} showPagination={true} rootValue={rootValue} value={isNil(value) ? defaultValue('array', undefined, nullable) : value}/>
        );
      }

      return (
        <Com {...this.props} showPagination={true} rootValue={rootValue} value={isNil(value) ? defaultValue(type, relation, nullable) : value} />
      );
    }
  };
}

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

function SpinWrapper({
  isFetching,
  children,
  ...props
}: {
  isFetching: boolean,
  children: React.Element<*>
}): React.Element<*> {
  return (
    <Spin indicator={antIcon} spinning={isFetching}>
      {React.cloneElement(children, props)}
    </Spin>
  )
}

export function shouldUpdate(value: any, newValue: any) {
  return value != newValue;
}
