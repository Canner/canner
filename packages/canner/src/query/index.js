// @flow

import {schemaToQueriesObject, objectToQueries} from './utils';
import {get, set, mapKeys} from 'lodash';

export class Query {
  schema: Object
  queries: Object
  variables: Object

  constructor({schema}: {schema: Object}) {
    this.schema = schema;
    const {queriesObj, variables} = schemaToQueriesObject(schema);
    this.variables = variables;
    this.queries = queriesObj;
  }

  updateQueries = (pathArr: Array<string>, field: string, value: any) => {
    const path = `${pathArr.join('.fields.')}.${field}`;
    if (field === 'args') {
      const args = get(this.queries, path);
      this.variables = {...this.variables, ...Object.keys(args).reduce((result: Object, key: string) => {
        result[args[key]] = value[key];

        return result;
      }, {})};
    } else {
      set(this.queries, path, value);
    }
  }

  getQueries = (pathArr: Array<string>): Object => {
    if (!pathArr || pathArr.length === 0) {
      return this.queries;
    }
    const path = pathArr.join('.fields.');
    return get(this.queries, path);
  }

  toGQL = (key?: string): string => {
    const variables = this.getVairables();
    if (key) {
      const obj = this.getQueries([key]);
      return objectToQueries({[key]: obj}, !obj.declareArgs, variables);  
    } else {
      return objectToQueries(this.queries, false, variables);
    }
  }

  getQueryKey = (key: string): string => {
    // return the string which includes the information of this query excluding the fields and declareArgs
    // e.g: postsConnection(first: $postsFirst,where: $postsWhere)
    const queries = this.getQueries([key]);
    delete queries.fields;
    delete queries.declareArgs;
    return objectToQueries({[key]: queries}, false)
  }

  getVairables = () => {
    return mapKeys(this.variables, (value, key) => key.substr(1));
  }
}