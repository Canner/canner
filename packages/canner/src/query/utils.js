// @flow

import pluralize from 'pluralize';
import lowerFirst from 'lodash/lowerFirst';
import upperFirst from 'lodash/upperFirst';
import {merge, mapValues, set} from 'lodash';
import {createSchema} from './schema/utils';
import {types} from './schema/types';

const DEFAULT_FIRST = 10;
const MAX = 999999999;

export function fieldToQueriesObject(field: any): any {
  let queriesObj = {};
  const variables = {};
  let variableTypes = {};
  const type = field.getType();
  switch (type) {
    case types.OBJECT: {
      field.forEach(childField => {
        const qlo = fieldToQueriesObject(childField);
        set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
      });
      break;
    }
    case types.ARRAY: {
      field.forEach(childField => {
        const qlo = fieldToQueriesObject(childField);
        set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
        merge(variables, qlo.variables);
        merge(variableTypes, qlo.variableTypes);
      });
      if (field.isEntity) {
        const {args, firstKey, afterKey, lastKey, beforeKey, whereKey, orderByKey} = genQuery();
        queriesObj.declareArgs = {
          ...variableTypes,
          [firstKey]: 'Int',
          [afterKey]: 'String',
          [lastKey]: 'Int',
          [beforeKey]: 'String',
          [whereKey]: `${typeKey(field.getKey())}WhereInput`,
          [orderByKey]: `${typeKey(field.getKey())}OrderByInput`
        };
        queriesObj.args = args;
        queriesObj.isPlural = true;
        queriesObj.connection = true;
        queriesObj.alias = field.getKey();
        set(queriesObj, ['fields', 'id'], null);
        variables[firstKey] = MAX;
        variables[whereKey] = {};
        const toolbar = field.getAttr('toolbar');
        const asyncToolbar = toolbar && toolbar.async;
        const defaultSort = toolbar && toolbar.sort && toolbar.sort.defaultSort;
        const permanentFilter = toolbar && toolbar.filter && toolbar.filter.permanentFilter;
        if (asyncToolbar) {
          variables[firstKey] = DEFAULT_FIRST;
        }
        if (permanentFilter) {
          variables[whereKey] = permanentFilter;
        }
        if (defaultSort) {
          variables[orderByKey] = `${defaultSort}_ASC`
        }
        
      }
      break;
    }

    case types.RELATION: {
      set(queriesObj, ['fields', 'id'], null);
      field.forEach(childField => {
        if (childField.getType() !== types.RELATION) {
          const qlo = fieldToQueriesObject(childField);
          set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
          merge(variables, qlo.variables);
          merge(variableTypes, qlo.variableTypes);
        }
      });
      // for now, fetch all toMany data
      // if (field.isToMany()) {
      //   const {args, firstKey, afterKey, lastKey, beforeKey, whereKey, orderByKey} = genQuery();
      //   queriesObj.args = args;
      //   queriesObj.isPlural = true;
      //   variables[firstKey] = defaultFirst;
      //   variableTypes = {
      //     ...variableTypes,
      //     [firstKey]: 'Int',
      //     [afterKey]: 'String',
      //     [lastKey]: 'Int',
      //     [beforeKey]: 'String',
      //     [whereKey]: `${typeKey(field.relationTo())}WhereUniqueInput`,
      //     [orderByKey]: `${typeKey(field.relationTo())}WhereUniqueInput`
      //   }
      // }
      break;
    }

    case types.DATETIME:
    case types.FILE:
    case types.GEOPOINT:
    case types.IMAGE: {
      field.forEach(childField => {
        const qlo = fieldToQueriesObject(childField);
        set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
      });
      break;
    }
    case types.BOOLEAN:
    case types.NUMBER:
    case types.INT:
    case types.ID:
    case types.STRING:
    default:
      queriesObj = null;
      break;
  }
  return {
    queriesObj,
    variables,
    variableTypes
  };
}

export function genQuery() {
  const firstKey = randomKey();
  const afterKey = randomKey();
  const lastKey = randomKey();
  const beforeKey = randomKey();
  const whereKey = randomKey();
  const orderByKey = randomKey();
  const args = {
    first: firstKey,
    after: afterKey,
    last: lastKey,
    before: beforeKey,
    where: whereKey,
    orderBy: orderByKey
  };
  return {args, firstKey, afterKey, lastKey, beforeKey, whereKey, orderByKey};
}

export function schemaToQueriesObject(schema: any) {
  const rootFields = createSchema(schema);
  const rootVariables = {};
  let rootVariableTypes = {};
  const queriesObj = mapValues(rootFields, (v, key) => {
    const {variables, queriesObj, variableTypes} = fieldToQueriesObject(v);
    merge(rootVariables, variables);
    rootVariableTypes[key] = variableTypes;
    return queriesObj;
  });

  return {
    queriesObj,
    variables: rootVariables
  }
}

export function objectToQueries(o: Object, close: boolean = true, variables?: Object) {

  const result = Object.keys(o).map(key => {
    let query = `${key}`;
    let element = o[key];
    if (!element) {
      return `${query}`;
    }

    if (element.declareArgs) {
      const originElement = {...element};
      const args = originElement.declareArgs;
      delete originElement.declareArgs;
      query = 'query';
      element = {
        args,
        fields: {
          [key]: originElement
        }
      };
      key = 'query';
    }

    if (element.isPlural) {
      query = pluralize.plural(lowerFirst(query));
    }

    if (element.connection) {
      query = `${query}Connection`;
    }

    if (element.alias) {
      query = `${element.alias}: ${query}`;
    }

    if (element.args) {
      if (key === 'query') {
        const args = genDeclareArgs(element.args, variables);
        query = args ? `${query}(${args})` : `${query}`;
      } else {
        const args = genArgs(element.args, variables);
        query = args ? `${query}(${args})` : `${query}`;
      }
    }

    if (element.fields) {
      let {fields} = element;
      if (element.connection) {
        fields = {
          edges: {
            fields: {
              cursor: null,
              node: {
                fields
              }
            }
          },
          pageInfo: {
            fields: {
              hasNextPage: null,
              hasPreviousPage: null
            }
          }
        }
      }
      fields = objectToQueries(fields, true, variables);
      
      query = `${query}${fields}`;
    }
    return `${query}`;
  }).join(' ');
  return close ? `{${result}}` : result;
}

function genArgs(args, variables) {
  return Object.keys(args)
    .filter(key => {
      if (variables && variables[args[key].substr(1)] === undefined) {
        return false;
      }
      return true;
    })
    .map(key => {
      const argValue = args[key];
      return `${key}: ${argValue}`
    }).join(',');
}

function genDeclareArgs(args, variables) {
  return Object.keys(args)
    .filter(key => {
      if (variables && variables[key.substr(1)] === undefined) {
        return false;
      }
      return true;
    })
    .map(key => {
      const argValue = args[key];
      return `${key}: ${argValue}`
    }).join(',');
}

function randomKey () {
  if (process.env.NODE_ENV === 'test') {
    return `$RANDOM_KEY`;
  }
  return `$KEY${Math.random().toString(36).substr(2, 7)}`
}

function typeKey(key) {
  return upperFirst(pluralize.singular(key));
}
