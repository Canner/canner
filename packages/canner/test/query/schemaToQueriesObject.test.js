import {schemaToQueriesObject} from '../../src/query/utils';

describe('schema to queries object', () => {
  it('should works', () => {
    const schema = {
      posts: {
        keyName: 'posts',
        type: 'array',
        path: 'posts',
        items: {
          type: 'object',
          items: {
            title: {
              path: 'posts/title',
              keyName: 'title',
              type: 'string',
            },
            staredPosts: {
              path: 'posts/staredPosts',
              keyName: 'staredPosts',
              type: 'relation',
              relation: {
                type: 'toMany',
                to: 'posts'
              }
            },
            bestAuthor: {
              keyName: 'bestAuthor',
              path: 'posts/bestAuthor',
              type: 'relation',
              relation: {
                type: 'toOne',
                to: 'users',
                typename: 'User'
              }
            },
            images: {
              keyName: 'images',
              path: 'posts/images',
              type: 'array',
              items: {
                type: 'image'
              }
            },
            wildCard: {
              keyName: 'wildCard',
              path: 'posts/wildCard',
              type: 'component'
            }
          }
        }
      },
      users: {
        type: 'array',
        path: 'users',
        items: {
          type: 'object',
          items: {
            name: {
              path: 'users/name',
              type: 'string'
            },
            email: {
              path: 'users/email',
              type: 'string'
            },
            images: {
              path: 'users/images',
              type: 'array',
              items: {
                type: 'image'
              }
            },
            posts: {
              path: 'users/posts',
              type: 'relation',
              relation: {
                type: 'toMany',
                to: 'posts'
              }
            },
            variants: {
              path: 'users/variants',
              type: 'json',
              items: {
                options: {
                  type: 'array'
                },
                variants: {
                  type: 'array'
                }
              }
            }
          }
        }
      },
    };
    expect(schemaToQueriesObject(schema)).toMatchSnapshot();
  });

  it('should not changed to lower case', () => {
    const schema = {
      otherMenuItem: {
        keyName: 'otherMenuItem',
        type: 'array',
        path: 'otherMenuItem',
        items: {
          type: 'object',
          items: {
            title: {
              keyName: 'title',
              path: 'otherMenuItem/title',
              type: 'string',
            },
            pathName: {
              keyName: 'pathName',
              path: 'otherMenuItem/pathName',
              type: 'string',
            }
          }
        }
      },
    };
    expect(schemaToQueriesObject(schema)).toMatchSnapshot();
  });
});
