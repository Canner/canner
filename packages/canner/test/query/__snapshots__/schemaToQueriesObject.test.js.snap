// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema to queries object should not changed to lower case 1`] = `
Object {
  "queriesObj": Object {
    "otherMenuItem": Object {
      "alias": "otherMenuItem",
      "args": Object {
        "after": "$otherMenuItemAfter",
        "before": "$otherMenuItemBefore",
        "first": "$otherMenuItemFirst",
        "last": "$otherMenuItemLast",
        "orderBy": "$otherMenuItemOrderBy",
        "where": "$otherMenuItemWhere",
      },
      "connection": true,
      "declareArgs": Object {
        "$otherMenuItemAfter": "String",
        "$otherMenuItemBefore": "String",
        "$otherMenuItemFirst": "Int",
        "$otherMenuItemLast": "Int",
        "$otherMenuItemOrderBy": "OtherMenuItemOrderByInput",
        "$otherMenuItemWhere": "OtherMenuItemWhereInput",
      },
      "fields": Object {
        "id": null,
        "pathName": null,
        "title": null,
      },
      "isPlural": true,
    },
  },
  "variables": Object {
    "$otherMenuItemFirst": undefined,
    "$otherMenuItemWhere": Object {},
  },
}
`;

exports[`schema to queries object should works 1`] = `
Object {
  "queriesObj": Object {
    "posts": Object {
      "alias": "posts",
      "args": Object {
        "after": "$postsAfter",
        "before": "$postsBefore",
        "first": "$postsFirst",
        "last": "$postsLast",
        "orderBy": "$postsOrderBy",
        "where": "$postsWhere",
      },
      "connection": true,
      "declareArgs": Object {
        "$postsAfter": "String",
        "$postsBefore": "String",
        "$postsFirst": "Int",
        "$postsLast": "Int",
        "$postsOrderBy": "PostOrderByInput",
        "$postsWhere": "PostWhereInput",
      },
      "fields": Object {
        "bestAuthor": Object {
          "fields": Object {
            "email": null,
            "id": null,
            "images": Object {
              "fields": Object {
                "contentType": null,
                "name": null,
                "size": null,
                "url": null,
              },
            },
            "name": null,
            "variants": null,
          },
        },
        "id": null,
        "images": Object {
          "fields": Object {
            "contentType": null,
            "name": null,
            "size": null,
            "url": null,
          },
        },
        "staredPosts": Object {
          "fields": Object {
            "id": null,
            "images": Object {
              "fields": Object {
                "contentType": null,
                "name": null,
                "size": null,
                "url": null,
              },
            },
            "title": null,
            "wildCard": undefined,
          },
        },
        "title": null,
        "wildCard": undefined,
      },
      "isPlural": true,
    },
    "users": Object {
      "alias": "users",
      "args": Object {
        "after": "$usersAfter",
        "before": "$usersBefore",
        "first": "$usersFirst",
        "last": "$usersLast",
        "orderBy": "$usersOrderBy",
        "where": "$usersWhere",
      },
      "connection": true,
      "declareArgs": Object {
        "$usersAfter": "String",
        "$usersBefore": "String",
        "$usersFirst": "Int",
        "$usersLast": "Int",
        "$usersOrderBy": "UserOrderByInput",
        "$usersWhere": "UserWhereInput",
      },
      "fields": Object {
        "email": null,
        "id": null,
        "images": Object {
          "fields": Object {
            "contentType": null,
            "name": null,
            "size": null,
            "url": null,
          },
        },
        "name": null,
        "posts": Object {
          "fields": Object {
            "id": null,
            "images": Object {
              "fields": Object {
                "contentType": null,
                "name": null,
                "size": null,
                "url": null,
              },
            },
            "title": null,
            "wildCard": undefined,
          },
        },
        "variants": null,
      },
      "isPlural": true,
    },
  },
  "variables": Object {
    "$postsFirst": undefined,
    "$postsWhere": Object {},
    "$usersFirst": undefined,
    "$usersWhere": Object {},
  },
}
`;
